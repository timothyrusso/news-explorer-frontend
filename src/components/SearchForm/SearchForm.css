.search-form {
  display: flex;
  justify-content: center;
  align-items: center;
}

.search-form__input {
  border-radius: 100px;
  min-height: 64px;
  max-width: 608px;
  border: 1px solid #D1D2D6;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.38;
  color: #B6BCBF;
  padding: 0 24px;
  width: 100%;
  box-sizing: border-box;
}

textarea:focus, input:focus {
  outline: none;
}

.search-form__button {
  background-color: #2F71E5;
  box-shadow: 0px 5px 15px rgba(14, 26, 57, 0.2);
  border-radius: 100px;
  min-height: 64px;
  max-width: 168px;
  width: 100%;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 500;
  font-size: 18px;
  line-height: 1.33;
  color: #fff;
  padding-top: 0;
  margin-left: -168px;
  transition: all linear 0.3s;
}

.search-form__button:hover {
  background-color: #347EFF;
  cursor: pointer;
}

.search-form__button:focus {
  background-color: #2A65CC;
}

@media screen and (max-width: 768px) {
  .search-form__input {
    max-width: 452px;
    min-height: 56px;
  }

  .search-form__button {
    min-height: 56px;
    max-width: 160px;
    margin-left: -160px;
  }
}

@media screen and (max-width: 320px) {
  .search-form {
    flex-direction: column;
  }

  .search-form__input {
    max-width: 288px;
    margin-bottom: 16px;
  }

  .search-form__button {
    max-width: 288px;
    margin-left: 0;
  }
}
