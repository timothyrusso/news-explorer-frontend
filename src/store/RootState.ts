import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
