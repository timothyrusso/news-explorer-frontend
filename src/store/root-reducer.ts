import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
