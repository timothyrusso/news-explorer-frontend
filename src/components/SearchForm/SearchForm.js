import { useState } from "react";
import "./SearchForm.css";

const SearchForm = ({ startLoadingNews, activateSearch }) => {
  const [searchInput, setSearchInput] = useState("");
  const [placeholder, setPlaceholder] = useState(false);

  const handleSearchInputChange = (evt) => {
    setSearchInput(evt.target.value);
  };

  const handleSubmit = (evt) => {
    if (searchInput) {
      startLoadingNews();
      evt.preventDefault();
      activateSearch({ search: searchInput });
      setPlaceholder(false);
    } else {
      evt.preventDefault();
      setPlaceholder(true);
    }
  };

  return (
    <form className="search-form" onSubmit={handleSubmit}>
      <input
        type="text"
        className={`search-form__input ${
          placeholder ? "search-form__input_type_error" : ""
        }`}
        id="search-input"
        name="search"
        placeholder={!placeholder ? "Enter topic" : "Please enter a keyword"}
        value={searchInput}
        onChange={handleSearchInputChange}
      ></input>
      <button className="search-form__button">Search</button>
    </form>
  );
};

export default SearchForm;
