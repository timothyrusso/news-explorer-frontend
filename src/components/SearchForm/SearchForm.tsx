import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { setShowArticleFalseAction } from '../../store/article/article.actions';
import { setIsLoadingTrueAction } from '../../store/toggles/toggles.actions';
import { startpoint, newsPerPage } from '../../utils/constants';
import {
  setNextThreeArticlesToThreeAction,
  fetchArticlesAction,
  fetchAllArticlesAction,
  setShowArticleTrueAction,
} from '../../store/article/article.actions';
import {
  setGenericServerErrorFalseAction,
  setGenericServerErrorTrueAction,
} from '../../store/errors/errors.actions';
import { setSearchKeywordAction } from '../../store/user/user.actions';
import { setIsLoadingFalseAction } from '../../store/toggles/toggles.actions';
import { getNewsInfo } from '../../utils/api';
import './SearchForm.css';

const SearchForm = () => {
  const [searchInput, setSearchInput] = useState('');
  const [placeholder, setPlaceholder] = useState(false);

  const dispatch = useDispatch();

  const handleSearchInputChange = (
    evt: React.ChangeEvent<HTMLInputElement>
  ) => {
    setSearchInput(evt.target.value);
  };

  const activateSearch = (
    data: { search: string },
    start = startpoint,
    end = newsPerPage
  ) => {
    localStorage.removeItem('news');
    dispatch(setNextThreeArticlesToThreeAction());
    dispatch(setGenericServerErrorFalseAction());
    dispatch(setSearchKeywordAction(data.search));

    getNewsInfo({ search: data.search })
      .then((data) => {
        localStorage.setItem('news', JSON.stringify(data.articles));
        const newsFromStorage = localStorage.getItem('news');
        if (newsFromStorage !== null) {
          dispatch(
            fetchArticlesAction(JSON.parse(newsFromStorage).slice(start, end))
          );
          dispatch(fetchAllArticlesAction(JSON.parse(newsFromStorage)));
        }
      })
      .catch((err) => {
        console.log(err);
        dispatch(setGenericServerErrorTrueAction());
      })
      .finally(() => {
        dispatch(setIsLoadingFalseAction());
        dispatch(setShowArticleTrueAction());
      });
  };

  const handleSubmit = (evt: React.FormEvent<HTMLFormElement>) => {
    if (searchInput) {
      dispatch(setShowArticleFalseAction());
      dispatch(setIsLoadingTrueAction());
      evt.preventDefault();
      activateSearch({ search: searchInput });
      setPlaceholder(false);
    } else {
      evt.preventDefault();
      setPlaceholder(true);
    }
  };

  return (
    <form className="search-form" onSubmit={handleSubmit}>
      <input
        type="text"
        className={`search-form__input ${
          placeholder ? 'search-form__input_type_error' : ''
        }`}
        id="search-input"
        name="search"
        placeholder={!placeholder ? 'Enter topic' : 'Please enter a keyword'}
        value={searchInput}
        onChange={handleSearchInputChange}
      ></input>
      <button className="search-form__button">Search</button>
    </form>
  );
};

export default SearchForm;
