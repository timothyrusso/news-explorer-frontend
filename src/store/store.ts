import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
