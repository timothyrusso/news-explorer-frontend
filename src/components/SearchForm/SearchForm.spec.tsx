import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store/store';
import SearchForm from './SearchForm';

describe('SearchForm component', () => {
  it('Should render the SearchForm component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
