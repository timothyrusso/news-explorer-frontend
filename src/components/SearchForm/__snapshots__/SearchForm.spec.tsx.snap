// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm component Should render the SearchForm component 1`] = `
<DocumentFragment>
  <form
    class="search-form"
  >
    <input
      class="search-form__input "
      id="search-input"
      name="search"
      placeholder="Enter topic"
      type="text"
      value=""
    />
    <button
      class="search-form__button "
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
